### PR TITLE
COMPLETED - memory

### PR DESCRIPTION
**Pull request by MMDPROJECT**
Completed memory for storing transitions as tuples of form (currrent_state, action, reward, next_state, done). 
Notes:
1. The underlying data structure is list, and it keeps appending transitions until the number of stored transitions reach `max_num_transitions`, in which case, it starts to pop off elements which have been appended first.
2. Sampling of mini-batches is not allowed until the list has stored a sufficient number of transitions, i.e. `mini_batch_size`. We should check the size of the list before sampling via `get_num_of_curr_transitions`.

### CODE REVIEW

File: memory.py [hunk 1]
```diff
@@ -1,4 +1,26 @@
...
+
+        trans = (curr_state, action, reward, next_state, done)
+        self.lst.append(trans)
+    
+    def sample_mini_batch(self):
+        if len(self.lst) < self.mini_batch_size:
+            raise Exception('Don\'t try to sample mini-batches while number of stored transitions < mini_batch_size')
+        
+        idxs = np.random.randint(0, len(self.lst), self.mini_batch_size)
+        batch = (self.lst[idx] for idx in idxs)
```
Comment (Retriever5500) [inline memory.py:25]: This isn't how we want to sample from out transitions. What this does is it makes a tuple of tuples of single transitions, what we want is batch them together. For example, for observations the dimensions are (4, 84, 84). What we want in return from sample is (32, 4, 84, 84) for example. (32 is mini-batch size).
So your return should be a tuple of tensors, with an extra dimension of the batch size.

**Review by Retriever5500 — CHANGES_REQUESTED** [review]

File: memory.py (hunk 1 above)
Comment (MMDPROJECT) [reply memory.py:25]: So you mean I have to stack all the data of all of the samples together across all of the different items we have (curr_state, action, reward, etc.) and make a tensor for each that contains the data for all of the samples right ?
e.g. if rewards are as follows: curr_state_1 = [0, 0], curr_state_ 2 = [1, 1], curr_state_3 = [2, 2], when we extract the first element of the batch it should looks like this [[0, 0], [1, 1], [2, 2]] ? 

If that's the case I will fix it now.

File: memory.py [hunk 2]
```diff
@@ -1,4 +1,31 @@
...
+    def sample_mini_batch(self):
+        if len(self.lst) < self.mini_batch_size:
+            raise Exception('Don\'t try to sample mini-batches while number of stored transitions < mini_batch_size')
+        
+        idxs = np.random.randint(0, len(self.lst), self.mini_batch_size)
+        
+        samples = (self.lst[idx] for idx in idxs)
+        
+        # changed the name to mini_batch in order to make the namings for similar to the workflow we use
+        mini_batch = tuple(map(torch.tensor, zip(samples)))
```
Comment (Retriever5500) [inline memory.py:30]: No that's not it I'm afraid, but let's go on for now when the time comes to use the samples I think you'll get the idea of what should be done
